src/ed25519.rs

```diff
@@ -0,0 +1,117 @@
+use crate::Commit;
+use bigint::U256;
+use curve25519_dalek::edwards::{CompressedEdwardsY, EdwardsPoint};
+
+pub use curve25519_dalek::{constants::ED25519_BASEPOINT_POINT as H, scalar::Scalar};
+
+pub type Point = EdwardsPoint;
+
+lazy_static::lazy_static! {
+    /// Alternate generator of ed25519.
+    pub static ref H_PRIME: EdwardsPoint = {
+        CompressedEdwardsY(hex_literal::hex!(
+            "8b655970153799af2aeadc9ff1add0ea6c7251d54154cfa92c173a0dd39c1f94"
+        ))
+        .decompress()
+        .expect("edwards point")
+    };
+}
+
+pub struct PedersenCommitment(Point);
+
+impl PedersenCommitment {
+    /// Generate a Pedersen Commitment for the scalar `x`.
+    pub fn new<R: rand::RngCore + rand::CryptoRng>(rng: &mut R, x: Scalar) -> (Self, Scalar) {
+        let s = Scalar::random(rng);
+        let C_H = x * H + s * *H_PRIME;
+
+        (Self(C_H), s)
+    }
+}
+
+impl Commit for PedersenCommitment {
+    type Commitment = Point;
+    type Blinder = Scalar;
+
+    fn commit<R: rand::RngCore + rand::CryptoRng>(
+        rng: &mut R,
+        bit: bool,
+    ) -> (Self::Commitment, Self::Blinder) {
+        let b = bit_as_scalar(bit);
+        let (PedersenCommitment(C_H), s) = PedersenCommitment::new(rng, b);
+
+        (C_H, s)
+    }
+}
+
+/// Transform a bit into a `ed25519::Scalar`.
+pub fn bit_as_scalar(bit: bool) -> Scalar {
+    if bit {
+        Scalar::one()
+    } else {
+        Scalar::zero()
+    }
+}
+
+/// Calculate sum of `r * 2^i`, where `i` is the bit index.
+pub fn blinder_sum(s_is: Vec<Scalar>) -> Scalar {
+    let two = U256::from(2u8);
+    s_is.iter().enumerate().fold(Scalar::zero(), |acc, (i, s)| {
+        let exp = two.pow(U256::from(i));
```

Comment:
There is no need for bigint to compute powers of two sequentially.

Reply:
I didn't see a simple way around this because I don't think I can simply raise a `secp256k1::Scalar` or a `ed25519::Scalar` to a power.

I guess I could define my own `pow` functions for that.

Reply:
This turned out to be another good tip! It works really nicely with `ecdsa_fun`, but unfortunately the analogous function for the `curve25519_dalek::Scalar` does not work. I still haven't figured out why :shrug:

Reply:
I mean just maintain the power of two for each iteration:

```
let mut pow_2 = s(1);
let two = s!(2);
for i in 0..n {
   let result = pow_2 * thing;
   ...
   pow_2 = pow_2 * two;
}
```

Reply:
`Or even simpler just set a bit in a [u8;32] and do let nth_power_of_two = Scalar::from_bytes(bytes_with_nth_bit_set).`

Haha that's ingenious.

Reply:
You can also use left shift to calculate powers of 2

In Rust seems its something like (fill in `N`) 
```
let power = Scalar(N);
let res = Scaler(1) << power
```